infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogCompaction.java

```diff
@@ -0,0 +1,54 @@
+package org.corfudb.infrastructure.log;
+
+import com.codahale.metrics.Counter;
+import com.codahale.metrics.MetricRegistry;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.runtime.CorfuRuntime;
+import org.corfudb.util.MetricsUtils;
+
+import java.util.concurrent.*;
+
+/**
+ * Scheduled log compaction manager.
+ * Execute log compaction safely (prevent scheduled executor corruption if the task throws an exception)
+ * https://stackoverflow.com/questions/6894595/scheduledexecutorservice-exception-handling
+ */
+@Slf4j
+public class StreamLogCompaction {
+    public static final String STREAM_COMPACT_METRIC = StreamLogCompaction.class.getName();
+
+    private final ThreadFactory threadFactory = new ThreadFactoryBuilder()
+            .setDaemon(true)
+            .setNameFormat("LogUnit-Maintenance-%d")
+            .build();
+
+    /**
+     * How many times log compaction executed
+     */
+    private final Counter gc = CorfuRuntime.getDefaultMetrics().counter(STREAM_COMPACT_METRIC);
+
+    /**
+     * A scheduler, which is used to schedule periodic garbage collection.
+     */
+    private final ScheduledExecutorService scheduler = Executors.newSingleThreadScheduledExecutor(threadFactory);
+    private final ScheduledFuture<?> compactor;
+
+    public StreamLogCompaction(StreamLog streamLog, long initialDelay, long period, TimeUnit timeUnit) {
+        Runnable task = () -> {
```

Comment:
Did we create a new class for this lambda that wraps the method call in a try catch ? If so why could we not just do it in the original class itself or have a generic utility function that returns runnable which has try/catch built into it. We will not have write a special class everywhere we need this ?

Reply:
I believe we need a class for that, reasons:

 - I don't want to pollute LogUnit business logic with utility logic.
 - I also want to measure how many times the compaction happened. I wouldn't like to put that measurement counter into LogUnit. 
 - It's easier to write a test if compaction is in a separate class.
 - follow Solid - one class one responsibility.

Reply:
When you call it utility logic then why not make it a utility so that other places can use it as well. 
As regards keeping this shallow class around, as long as you do not take this principle to the extreme it is ok in a few cases (since you also added a counter to justify this change).   This shallow class can also use the common utility that you can provide.

Reply:
So, the most important thing for me is to have a test for StreamLogCompaction, without that class it won't be easy to test the compaction logic. My logic is:
 - question: why do I want to have a shallow class for compaction? 
 - Well, I want to have a test for that compaction functional. 
 - Will it be easy to test if I leave that functional in LogUnitServer? answer: looks like no because LogUnitServer has many dependencies like BatchWriter, StreamLog and in addition, I need to provide ServerContext. So there are a lot of dependencies, do I want to provide all that dependencies just to test my compaction logic? My answer is no.

Other places can use StreamLogCompaction as well I don't see any problems with that, maybe you could give me an example when having this class is not convenient?

Reply:
I do not think I understand this completely. I was hoping that you would have a utility like this 
Runnable wrappedRunnable ( lambda ) {
   return new Runnable ( {
              run() {
                   try {
                       lambda
                   }catch {}

              })
   }
}

And your StreamLogCompaction can use that utility. I do not think StreamLogCompaction is a very generic utility that is going to have a lot of use cases other than the one you are using it for.